nvflare/app_opt/job_launcher/docker_launcher.py

```diff
@@ -0,0 +1,195 @@
+
+        workspace_obj: Workspace = fl_ctx.get_prop(FLContextKey.WORKSPACE_OBJECT)
+        job_id = job_meta.get(JobConstants.JOB_ID)
+        app_custom_folder = workspace_obj.get_app_custom_dir(job_id)
+
+        python_path = f"{app_custom_folder}:$PYTHONPATH" if app_custom_folder != "" else "$PYTHONPATH"
+        job_name, cmd = self.get_command(job_meta, fl_ctx)
+        command = f' /bin/bash -c "export PYTHONPATH={python_path};{cmd}"'
+        self.logger.info(f"Launch image:{job_image}, run command: {command}")
+
+        client = docker.from_env()
+        try:
+            container = client.containers.run(
+                job_image,
+                command=command,
+                name=job_name,
+                network=self.network,
+                detach=True,
+                # remove=True,
+                volumes={
+                    self.workspace: {
+                        "bind": self.mount_path,
+                        "mode": "rw",
+                    },
+                },
+                # ports=ports,  # Map container ports to host ports (optional)
+            )
+            self.logger.info(f"Launch the job in DockerJobLauncher using image: {job_image}")
+
+            handle = DockerJobHandle(container)
+            try:
+                if handle.enter_states([DOCKER_STATE.RUNNING], timeout=self.timeout):
+                    return handle
+                else:
+                    handle.terminate()
+                    return None
+            except:
+                handle.terminate()
+                return None
+
+        except docker.errors.ImageNotFound:
+            self.logger.error(f"Failed to launcher job: {job_id} in DockerJobLauncher. Image '{job_image}' not found.")
+            return None
+        except docker.errors.APIError as e:
+            self.logger.error(f"Error starting container: {e}")
+            return None
+
+    def handle_event(self, event_type: str, fl_ctx: FLContext):
+        if event_type == EventType.GET_JOB_LAUNCHER:
+            job_meta = fl_ctx.get_prop(FLContextKey.JOB_META)
+            job_image = extract_job_image(job_meta, fl_ctx.get_identity_name())
+            if job_image:
+                add_launcher(self, fl_ctx)
+
+    @abstractmethod
+    def get_command(self, job_meta, fl_ctx) -> (str, str):
+        """To generate the command to launcher the job in sub-process
+
+        Args:
+            fl_ctx: FLContext
+            job_meta: job launcher data
+
+        Returns:
+            (container name, launch command)
+
+        """
+        pass
+
+
+class ClientDockerJobLauncher(DockerJobLauncher, ClientProcessJobLauncher):
+    def get_command(self, job_meta, fl_ctx) -> (str, str):
+        job_id = job_meta.get(JobConstants.JOB_ID)
+        command = self.generate_client_command(job_meta, fl_ctx)
+
+        return f"client-{job_id}", command
+
+
+class ServerDockerJobLauncher(DockerJobLauncher, ServerProcessJobLauncher):
```

Comment:
Should ServerDockerJobLauncher be an instance of ServerProcessJobLauncher?
They seems like different concepts?

Reply:
Use the job_launcher_utils to remove the inheritance relationship.